Update mongodb dependency to 5.0.1

alexisGandar opened the pull request:
Spring boot 3.3.X update use org.mongodb 5.0.1

This PR fix a breaking change related to mongodb 5.X.X : https://www.mongodb.com/docs/drivers/java/sync/v5.0/whats-new/#std-label-version-5.0
![image](https://github.com/mongounit/mongounit/assets/27289886/9db60183-8206-45f2-9690-96693096efa6)
The com.mongodb.client.MongoDatabase.listCollectionNames() return type changed. This PR adapt the MongoUnitUtil code to this change.

alexisGandar commented:
@ychaikin Do you still maintain this project ?

ychaikin commented:
> @ychaikin Do you still maintain this project ?

Yes, I do and thank you for the PR! I am planning to test this out and then cut a new release this weekend.

Thank you!

ychaikin commented:
@alexisGandar What version of Spring Boot are you using with this? I am a bit confused since the PR doesn't update the Spring Boot version yet you are saying that 3.3.x needs this change. The `pom.xml` still says ` <spring.version>2.3.4.RELEASE</spring.version>`

Can you elaborate?

Also, what version of JDK are you running this on?

alexisGandar commented:
I do use Spring Boot 3.3 in my project. Upgrading to this version caused the issue because the subdependency of mongo.org got upgraded to 5.x. In this PR I fixed the issue by upgrading mongo.org. I did not upgrade spring boot because it was not needed but it might be more logical to do that. I will look into it on Monday.
Thank you for the plugin !

ychaikin commented:
Right... I don't think just upgrading the Java class will be enough since the library won't compile with the current Spring Boot versions. I think what we need to do is find out the FIRST version of Spring Boot when it started using the new API and set the Spring Boot version in mongounit to that.

ychaikin commented:
After a quick search, it looks like Spring Boot 3.0 is the version that started using the new MongoDB driver. So, I am not sure if that actually means that MongoUnit needs to depend on Spring Boot 3.0 or if just updating the mongodb driver dependency is enough. Definitely need to update the mongodb driver dependency, but might have to update the Spring Boot version as well (which is basically transitively means updating the Spring Data version).

alexisGandar commented:
The version where the MongoDB driver has been updated seems to be Spring Boot 3.3 : 

- Spring Boot dep on 3.2 : https://mvnrepository.com/artifact/org.springframework.boot/spring-boot-dependencies/3.2.6
<img width="1227" alt="image" src="https://github.com/mongounit/mongounit/assets/27289886/5a4c2b3d-4e38-49a4-b842-6aee99a98ae5">

- Spring Boot dep on 3.3 : https://mvnrepository.com/artifact/org.springframework.boot/spring-boot-dependencies/3.3.0
<img width="1176" alt="image" src="https://github.com/mongounit/mongounit/assets/27289886/62c80c62-b449-4007-b5c0-18eed8f4643d">

I updated the PR to Spring Boot 3.3.
In my opinion we should leave the dependency in spring boot as it will avoid importing different version of the dep for nothing.

I also updated Java to JDK 17 as it is the minimum required version for Spring Boot 3.X

ychaikin commented:
"In my opinion we should leave the dependency in spring boot as it will avoid importing different version of the dep for nothing."

Sorry, not 100% following what you are trying to say here. Could you rephrase?

alexisGandar commented:
"I am not sure if that actually means that MongoUnit needs to depend on Spring Boot 3.0 or if just updating the mongodb driver dependency is enough" 
Sorry I may have not understand your previous message. What I was trying to say is that I removed the mongodb-driver-sync dependency from the pom.xml as it's already included in the dependencies of Spring Boot 3.3 with the good version (5.0.1)

alexisGandar commented:
It's a dependency I added before the upgrade to Spring Boot 3.3 (it was not in the pom before this PR). Look at my commit "Update Spring Boot to 3.3.0" I deleted a dependency that was added in the first commit "Update mongodb dependency to 5.0.1". The reason beeing that the dependency is now included in the Spring Boot sub-dependencies in 3.3 so there is no need to "force" the version in the pom.

ychaikin commented:
Ok, I see. I will try to take a look and test things out this weekend. Hopefully, we should be good to go after that.

Thank you!!

ychaikin commented:
@alexisGandar Sorry for the delay. I need a few more days to get to this.

ychaikin commented on src/main/java/org/mongounit/MongoUnitUtil.java:
```diff
@@ -176,12 +176,11 @@ public static void toDatabase(
     // Iterate over all collections in the db and drop them
-    for (String collectionName : mongoDatabase.listCollectionNames()) {
-
+    mongoDatabase.listCollectionNames().forEach(collectionName -> {
```
@alexisGandar I am assuming this is just a newer syntax cleanup? This really doesn't have anything to do with the version update, does it?

ychaikin commented on src/main/java/org/mongounit/MongoUnitUtil.java:
```diff
@@ -261,13 +260,9 @@ private static List<String> getCollectionNamesToUse(
-    }
-
+    List<String> collectionNames = new ArrayList<>();
+    mongoDatabase.listCollectionNames().forEach(collectionNames::add);
```
@alexisGandar  Same question here. Just using newer syntax, but doesn't have to do with the intention of the PR, correct?

ychaikin commented on src/main/java/org/mongounit/MongoUnitUtil.java:
```diff
@@ -176,12 +176,11 @@ public static void toDatabase(
     // Iterate over all collections in the db and drop them
-    for (String collectionName : mongoDatabase.listCollectionNames()) {
-
+    mongoDatabase.listCollectionNames().forEach(collectionName -> {
```
@alexisGandar This update is just to use the newer syntax in Java, correct? I don't think this code was breaking with the version update, was it?

alexisGandar replied:
It's not related to the java version. It's related to the upgrade of MongoDB drivers. This is the reason of this PR. After upgrading spring boot to 3.3.0 the MongoDB driver sub-dependency were also updated and in the update they changed the return type of the listCollectionNames. This change was required to avoid an error.

ychaikin commented on src/main/java/org/mongounit/MongoUnitUtil.java:
```diff
@@ -261,13 +260,9 @@ private static List<String> getCollectionNamesToUse(
-    }
-
+    List<String> collectionNames = new ArrayList<>();
+    mongoDatabase.listCollectionNames().forEach(collectionNames::add);
```
@alexisGandar Same question here... This update is just to use the newer syntax in Java, correct? I don't think this code was breaking with the version update, was it?

alexisGandar replied:
Same comment thant above => It's not related to the java version. It's related to the upgrade of MongoDB drivers. This is the reason of this PR. After upgrading spring boot to 3.3.0 the MongoDB driver sub-dependency were also updated and in the update they changed the return type of the listCollectionNames. This change was required to avoid an error.

ychaikin replied:
Strange that you'd get an error since `public interface ListCollectionNamesIterable extends MongoIterable<String>` so it extends `MongoIterable<String>`.

Either way, it's update to the syntax, so whatever. Merging.

ychaikin commented:
Closes https://github.com/mongounit/mongounit/issues/37

ychaikin commented:
@alexisGandar FYI: New release (with a fix of timezone for seeding dates) is now on Maven central: https://repo1.maven.org/maven2/org/mongounit/mongounit/3.0.0/

Thanks for the help! (https://github.com/mongounit/mongounit/releases)
